feat: runtime enforcer rules (#425)
kubewarden/policies: policies/runtime-enforcer-policy/README.md

[![Kubewarden Policy Repository](https://github.com/kubewarden/community/blob/main/badges/kubewarden-policies.svg)](https://github.com/kubewarden/community/blob/main/REPOSITORIES.md#policy-scope)
[![Stable](https://img.shields.io/badge/status-stable-brightgreen?style=for-the-badge)](https://github.com/kubewarden/community/blob/main/REPOSITORIES.md#stable)

> ## Runtime-enforcer policy

Reviewer: We avoid to repeat the policy name because the policy description is then embedded into UIs and this repetition makes things a bit ugly

```suggestion
```

This policy is designed to enforce constraints on runtime enforcer. This
enforces that workload can only be deployed when its runtime enforcer

> policy specified via "security.rancher.io/policy" label.

Reviewer: ```suggestion
This policy ensures Kubernetes workloads that make use of Runtime Enforcer
to protect themselves are referring an existing `WorkloadPolicy`.
```

## Configuration

Currently there is no configurable options at this time.

## Behavior

This policy checks the container label of supported workloads to see if
a runtime-enforcer policy is specified:

- Pod
- ReplicaSet
- Deployment
- StatefulSet
- DaemonSet
- Job

> - CronJob

Reviewer: ```suggestion
Kubernetes Pods can be protected by Runtime Enforcer by adding the
`security.rancher.io/policy` label to them. This policy ensures that
the `WorkloadPolicy` referenced by that label exists.

The policy can validate Pods and higher order Kubernetes workload types,
such as:

- CronJob
- DaemonSet
- Deployment
- Job
- ReplicaSet
- StatefulSet
```

## Deployment

This policy requires extra permission to read WorkloadPolicy resource of
runtime-enforcer.  Before deploying this policy, update the policy

> server's permission to include security.rancher.io workloadpolicies. 

Reviewer: ```suggestion
This policy requires extra permission to read `WorkloadPolicy` resource
defined by Runtime Enforcer.

The Service Account used by the Policy Server hosting this policy
must be granted `GET`, `LIST` and `WATCH` privileges to the
`WorkloadPolicy` resource:

This can be achieved by creating the following RBAC resources:

```yaml
apiVersion: rbac.authorization.k8s.io/v1
kind: ClusterRole
metadata:
  name: runtime-enfocer-workloadpolicy-viewer
rules:
  - apiGroups:
      - security.rancher.io
    resources:
      - workloadpolicies
    verbs:
      - get
      - list
      - watch
---
apiVersion: rbac.authorization.k8s.io/v1
kind: ClusterRoleBinding
metadata:
  name: runtime-enforcer-workloadpolicy-viewer-policy-server
roleRef:
  apiGroup: rbac.authorization.k8s.io
  kind: ClusterRole
  name: runtime-enfocer-workloadpolicy-viewer
subjects:
  - kind: ServiceAccount
    name: policy-server
    namespace: kubewarden
```
```

```
policyServer:
  permissions:
    # Make sure that you keep the existing permissions, then add the
    # below resource:
    - apiGroup: "security.rancher.io"
      resources:
        - workloadpolicies
```

Deploy the policy as a `ClusterAdmissionPolicy` with the `contextAwareResources`

> field properly set. Use the following command to scaffold the policy:

Reviewer: ```suggestion
This is a context aware policy, hence it can be deployed only as a
`ClusterAdmissionPolicy`.

Use the following command to scaffold the policy:
```

Reviewer 2: I would specify:

```suggestion
This is a Kubernetes context aware policy, hence it can be deployed only as a
`ClusterAdmissionPolicy`.

Use the following command to scaffold the policy:
```

[![Kubewarden Policy Repository](https://github.com/kubewarden/community/blob/main/badges/kubewarden-policies.svg)](https://github.com/kubewarden/community/blob/main/REPOSITORIES.md#policy-scope)
[![Stable](https://img.shields.io/badge/status-stable-brightgreen?style=for-the-badge)](https://github.com/kubewarden/community/blob/main/REPOSITORIES.md#stable)

> This policy ensures Kubernetes workloads that make use of Runtime Enforcer

Reviewer: ```suggestion
This policy ensures Kubernetes workloads that make use of
[Runtime Enforcer](https://github.com/rancher-sandbox/runtime-enforcer)
```

`security.rancher.io/policy` label to them. This policy ensures that
the `WorkloadPolicy` referenced by that label exists.
The policy can validate Pods and higher order Kubernetes workload types,
such as:

- CronJob
- DaemonSet
- Deployment
- Job
- ReplicaSet
- StatefulSet

## Deployment

This policy requires extra permission to read `WorkloadPolicy` resource
defined by Runtime Enforcer.

The Service Account used by the Policy Server hosting this policy
must be granted `GET`, `LIST` and `WATCH` privileges to the
`WorkloadPolicy` resource:

This can be achieved by creating the following RBAC resources:

```yaml
apiVersion: rbac.authorization.k8s.io/v1
kind: ClusterRole
metadata:
  name: runtime-enforcer-workloadpolicy-viewer
rules:
  - apiGroups:
      - security.rancher.io
    resources:
      - workloadpolicies
    verbs:
      - get
      - list
      - watch
---
apiVersion: rbac.authorization.k8s.io/v1
kind: ClusterRoleBinding
metadata:
  name: runtime-enforcer-workloadpolicy-viewer-policy-server
roleRef:
  apiGroup: rbac.authorization.k8s.io
  kind: ClusterRole
  name: runtime-enforcer-workloadpolicy-viewer
subjects:
  - kind: ServiceAccount
    name: policy-server

>     namespace: kubewarden

Reviewer 3: Thanks for providing this in the README.md! :heart:

The policy can validate Pods and higher order Kubernetes workload types,
such as:

- CronJob
- DaemonSet
- Deployment
- Job
- ReplicaSet
- StatefulSet

## Deployment

This policy requires extra permission to read `WorkloadPolicy` resource
defined by Runtime Enforcer.

The Service Account used by the Policy Server hosting this policy
must be granted `GET`, `LIST` and `WATCH` privileges to the
`WorkloadPolicy` resource:

This can be achieved by creating the following RBAC resources:

```yaml
apiVersion: rbac.authorization.k8s.io/v1
kind: ClusterRole
metadata:
  name: runtime-enforcer-workloadpolicy-viewer
rules:
  - apiGroups:
      - security.rancher.io
    resources:
      - workloadpolicies
    verbs:
      - get
      - list
      - watch
---
apiVersion: rbac.authorization.k8s.io/v1
kind: ClusterRoleBinding
metadata:
  name: runtime-enforcer-workloadpolicy-viewer-policy-server
roleRef:
  apiGroup: rbac.authorization.k8s.io
  kind: ClusterRole
  name: runtime-enforcer-workloadpolicy-viewer
subjects:
  - kind: ServiceAccount
    name: policy-server
    namespace: kubewarden
```

This is a context aware policy, hence it can be deployed only as a
`ClusterAdmissionPolicy`.

> Use the following command to scaffold the policy:

Reviewer 2: this could be removed if we are publishing to artifacthub.io, we automatically publish there with [install metadata](https://github.com/kubewarden/policies/blob/artifacthub/policies/cel-policy/artifacthub-pkg.yml#L27-L34) that is automatically generated.

See: https://artifacthub.io/packages/kubewarden/kubewarden-policy-library/cel-policy?modal=install
